torchmultimodal/modules/encoders/text_encoder.py

```diff
@@ -0,0 +1,155 @@
+    TransformerEncoder,
+    TransformerOutput,
+)
+from torchmultimodal.utils.attention import get_extended_attention_mask
+
+
+class TextEncoder(nn.Module):
+    """
+    General text transformer encoder with embeddings, similar to BERT / RoBERTa.
+    Can be constructed with any user-provided embeddings and encoder.
+
+    Based on https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/modeling_bert.py#L870
+
+    Attributes:
+        embeddings (nn.Module): Module that projects text token ids into embeddings.
+            ``forward()`` should follow interface:
+                input_ids: Optional[Tensor],
+                token_type_ids: Optional[Tensor],
+                position_ids: Optional[Tensor],
+                inputs_embeds: Optional[Tensor],
+        encoder (nn.Module): Module for transformer encoder. ``forward()`` should follow interface:
+            Inputs:
+                hidden_states: Tensor,
+                attention_mask: Optional[Tensor], shape [batch_size, num_heads, query_seq_length, key_seq_length]
+            Returns:
+                ``TransformerOutput``
+        layernorm (nn.Module, optional): Module for layernorm to be applied after encoder, if provided.
+        pooler (nn.Module, optional): Module for pooler to be applied after layernorm, if provided.
+
+    Args:
+        input_ids (Tensor, optional): Tensor of input vocab token ids of shape [batch, seq_len].
+        attention_mask (Tensor, optional): Tensor indicating which tokens to attend to, shape [batch, seq_len]
+        token_type_ids (Tensor, optional): Tensor of input token type ids of shape [batch, seq_len]. In BERT,
+            used to indicate whether a word is in sentence A or B for next sentence prediction
+        position_ids (Tensor, optional): Tensor of input position ids of shape [batch, seq_len]
+        inputs_embeds (Tensor, optional): Tensor of input embeddings of shape [batch, hidden_size],
+            if embeddings are calculated elsewhere
+
+    Raises:
+        ValueError: if input_ids and inputs_embeds are both None
+    """
+
+    def __init__(
+        self,
+        embeddings: nn.Module,
+        encoder: nn.Module,
+        layernorm: Optional[nn.Module] = None,
+        pooler: Optional[nn.Module] = None,
+        weight_init_fn: Optional[Callable] = None,
+    ) -> None:
+        super().__init__()
+
+        self.embeddings = embeddings
+        self.encoder = encoder
+        self.layernorm = layernorm
+        self.pooler = pooler
+
+        if weight_init_fn:
+            self.apply(weight_init_fn)
+
+    def forward(
+        self,
+        input_ids: Optional[Tensor] = None,
+        attention_mask: Optional[Tensor] = None,
+        token_type_ids: Optional[Tensor] = None,
+        position_ids: Optional[Tensor] = None,
+        inputs_embeds: Optional[Tensor] = None,
+    ) -> TransformerOutput:
+        if input_ids is not None:
```

Comment:
Noob question: when would input IDs be none? Also the logic to get device is the same as in the text embeddings class, maybe worth extracting to a common util to reduce boilerplate?

Reply:
there are cases where you may want to use your own word embeddings but still want the token_type and pos embeddings from this module, in which case the user can pass in `inputs_embeds` instead of `input_ids`. This is similar to HuggingFace's [design](https://github.com/huggingface/transformers/blob/d0acc9537829e7d067edbb791473bbceb2ecf056/src/transformers/models/bert/modeling_bert.py#L210)